File: t/CS/ImageProcessingDemo/UI/Command/DeskewCheckDialog.cs
// *************************************************************
// Copyright (c) 1991-2019 LEAD Technologies, Inc.              
// All Rights Reserved.                                         
// *************************************************************
using System;
using System.Collections.Generic;
using System.ComponentModel;
using System.Data;
using System.Drawing;
using System.Text;
using System.Windows.Forms;

using Leadtools;
using Leadtools.ImageProcessing;
using Leadtools.ImageProcessing.Core;

namespace ImageProcessingDemo
{
   //This dialog is used to allow the user to specify values which
   //will be used for the DeskewCheckCommand

   public partial class DeskewCheckDialog : Form
   {
      private DeskewCheckCommand _DeskewCheckCommand = null;
      private RasterColor _Color;
      private RasterImage _image;
      public bool ReturnAngleOnly = false;

      public DeskewCheckDialog(RasterImage image)
      {
         _image = image;
         InitializeComponent();
         _DeskewCheckCommand = new DeskewCheckCommand();
         _Color = Leadtools.Demos.Converters.FromGdiPlusColor(Color.Black);
         _rbFast.Enabled = (_image.BitsPerPixel == 1);

         //Set command default values
         InitializeUI();
      }
      public DeskewCheckDialog(DeskewCheckCommand DeskewCheckCommand)
      {
         InitializeComponent();
         _DeskewCheckCommand = DeskewCheckCommand;
         InitializeUI();
      }
      public DeskewCheckCommand DeskewCheckCommand
      {
         get
         {
            //Update command values
            UpdateCommand();
            return _DeskewCheckCommand;
         }
         set
         {
            _DeskewCheckCommand = value;
            InitializeUI();
         }
      }

      public RasterColor color
      {
         get
         {
            return _Color;
         }
         set
         {
            _Color = value;
         }
      }

      private void _btnOk_Click(object sender, EventArgs e)
      {
         UpdateCommand();
         this.DialogResult = DialogResult.OK;
         this.Close();
      }

      private void _btnCancel_Click(object sender, EventArgs e)
      {
         this.DialogResult = DialogResult.Cancel;
      }

      private void InitializeUI()
      {
         // Fill
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.FillExposedArea) == DeskewCheckCommandFlags.FillExposedArea)
         {
            _cbFillExposedArea.Checked = true;
            _lblColor.BackColor = Leadtools.Demos.Converters.ToGdiPlusColor(_DeskewCheckCommand.FillColor);
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(Color.Black);
            _lblColor.Enabled = true;
         }
         //Deskew Image
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.DeskewImage) == DeskewCheckCommandFlags.DeskewImage)
         {
            _rbDeskewImage.Checked = true;
         }

         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.ReturnAngleOnly) == DeskewCheckCommandFlags.ReturnAngleOnly)
         {
            _rbReturnAngleOnly.Checked = true;
         }
         // Threshold
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.Threshold) == DeskewCheckCommandFlags.Threshold)
         {
            _cbThreshold.Checked = true;
         }

         // Quality
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.RotateLinear) == DeskewCheckCommandFlags.RotateLinear)
         {
            _rbLow.Checked = true;
         }
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.RotateResample) == DeskewCheckCommandFlags.RotateResample)
         {
            _rbMedium.Checked = true;
         }
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.RotateBicubic) == DeskewCheckCommandFlags.RotateBicubic)
         {
            _rbHigh.Checked = true;
         }

         // Type
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.DocumentImage) == DeskewCheckCommandFlags.DocumentImage)
         {
            _rbTextOnly.Checked = true;
         }
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.DocumentAndPictures) == DeskewCheckCommandFlags.DocumentAndPictures)
         {
            _rbTextImages.Checked = true;
         }

         // Speed
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.UseNormalRotate) == DeskewCheckCommandFlags.UseNormalRotate)
         {
            _rbNormal.Checked = true;
         }
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.UseHighSpeedRotate) == DeskewCheckCommandFlags.UseHighSpeedRotate)
         {
            _rbFast.Checked = true;
         }

         //Algorithm
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.DoNotUseCheckDeskew) == DeskewCheckCommandFlags.DoNotUseCheckDeskew)
         {
            _rbOrdinaryDeskew.Checked = true;
         }
         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.UseNormalCheckDeskew) == DeskewCheckCommandFlags.UseNormalCheckDeskew)
         {
            _rbBankCheckDeskew.Checked = true;
         }

         if ((_DeskewCheckCommand.Flags & DeskewCheckCommandFlags.UseLineDetectionToDeskewCheck) == DeskewCheckCommandFlags.UseLineDetectionToDeskewCheck)
         {
            _rbLineDetectionDeskew.Checked = true;
         }
      }

      private void UpdateCommand()
      {
         //DeskewCheckCommand.FillColor = new RasterColor(Color.Red);// _Color;
         _DeskewCheckCommand.Flags = (DeskewCheckCommandFlags)0;

         // Fill
         if (_cbFillExposedArea.Checked)
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.FillExposedArea;
         }
         else
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.DoNotFillExposedArea;

         //Deskew Image
         if (_rbDeskewImage.Checked)
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.DeskewImage;
         else
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.ReturnAngleOnly;
            ReturnAngleOnly = true;
         }

         // Threshold
         if (_cbThreshold.Checked)
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.Threshold;
         else
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.NoThreshold;

         // Quality
         if (_rbLow.Checked)
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.RotateLinear;
         }
         else if (_rbMedium.Checked)
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.RotateResample;
         }
         else
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.RotateBicubic;
         }

         // Type
         if (_rbTextOnly.Checked)
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.DocumentImage;
         }
         else
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.DocumentAndPictures;
         }

         // Speed
         if (_rbNormal.Checked)
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.UseNormalRotate;
         }
         else
         {
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.UseHighSpeedRotate;
         }

         //Algorithm
         if (_rbOrdinaryDeskew.Checked)
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.DoNotUseCheckDeskew;
         if (_rbBankCheckDeskew.Checked)
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.UseNormalCheckDeskew;
         if (_rbLineDetectionDeskew.Checked)
            _DeskewCheckCommand.Flags |= DeskewCheckCommandFlags.UseLineDetectionToDeskewCheck;


      }

      private void _cbFillExposedArea_CheckedChanged(object sender, EventArgs e)
      {
         _lblColor.Enabled = _cbFillExposedArea.Checked;
      }

      private void lblColor_Click(object sender, EventArgs e)
      {
         ColorDialog dlg = new ColorDialog();
         dlg.Color = _lblColor.BackColor;
         if (dlg.ShowDialog() == DialogResult.OK)
         {
            _lblColor.BackColor = dlg.Color;
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(dlg.Color);
         }
      }

      private void _btnColor_Click(object sender, EventArgs e)
      {
         ColorDialog dlg = new ColorDialog();
         dlg.AllowFullOpen = true;
         dlg.AnyColor = true;
         DialogResult res = dlg.ShowDialog(this);
         if (res == DialogResult.OK)
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(dlg.Color);
         _lblColor.Refresh();
      }

      private void _lblColor_Paint(object sender, PaintEventArgs e)
      {
         e.Graphics.FillRectangle(new SolidBrush(Leadtools.Demos.Converters.ToGdiPlusColor(_Color)), _lblColor.ClientRectangle);
      }
   }
}

File: t/VB/ImageProcessingDemo/UI/Command/DeskewCheckDialog.vb
' *************************************************************
' Copyright (c) 1991-2019 LEAD Technologies, Inc.              
' All Rights Reserved.                                         
' *************************************************************
Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.ComponentModel
Imports System.Data
Imports System.Drawing
Imports System.Text
Imports System.Windows.Forms

Imports Leadtools
Imports Leadtools.ImageProcessing
Imports Leadtools.ImageProcessing.Core

Namespace ImageProcessingDemo
   'This dialog is used to allow the user to specify values which
   'will be used for the DeskewCheckCommand

   Partial Public Class DeskewCheckDialog : Inherits Form
      Private _DeskewCheckCommand As DeskewCheckCommand = Nothing
      Private _Color As RasterColor
      Public ReturnAngleOnly As Boolean = False
      Private _Image As RasterImage

      Public Sub New(ByVal Image As RasterImage)
         InitializeComponent()
         _Image = Image

         If Not (_Image.BitsPerPixel = 1) Then
            _rbFast.Enabled = False
         End If

         _DeskewCheckCommand = New DeskewCheckCommand()
         _Color = Leadtools.Demos.Converters.FromGdiPlusColor(System.Drawing.Color.Black)

         'Set command default values
         InitializeUI()
      End Sub
      Public Sub New(ByVal DeskewCheckCommand As DeskewCheckCommand)
         InitializeComponent()
         _DeskewCheckCommand = DeskewCheckCommand
         InitializeUI()
      End Sub
      Public Property DeskewCheckCommand() As DeskewCheckCommand
         Get
            'Update command values
            UpdateCommand()
            Return _DeskewCheckCommand
         End Get
         Set(ByVal value As DeskewCheckCommand)
            _DeskewCheckCommand = value
            InitializeUI()
         End Set
      End Property

      Public Property color() As RasterColor
         Get
            Return _Color
         End Get
         Set(ByVal value As RasterColor)
            _Color = value
         End Set
      End Property

      Private Sub _btnOk_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnOk.Click
         UpdateCommand()
         Me.DialogResult = DialogResult.OK
         Me.Close()
      End Sub

      Private Sub _btnCancel_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnCancel.Click
         Me.DialogResult = DialogResult.Cancel
      End Sub

      Private Sub InitializeUI()
         ' Fill
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.FillExposedArea) = DeskewCheckCommandFlags.FillExposedArea Then
            _cbFillExposedArea.Checked = True
            _lblColor.BackColor = Leadtools.Demos.Converters.ToGdiPlusColor(_DeskewCheckCommand.FillColor)
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(System.Drawing.Color.Black)
            _lblColor.Enabled = True
         End If
         'Deskew Image
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.DeskewImage) = DeskewCheckCommandFlags.DeskewImage Then
            _rbDeskewImage.Checked = True
         End If

         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.ReturnAngleOnly) = DeskewCheckCommandFlags.ReturnAngleOnly Then
            _rbReturnAngleOnly.Checked = True
         End If
         ' Threshold
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.Threshold) = DeskewCheckCommandFlags.Threshold Then
            _cbThreshold.Checked = True
         End If

         ' Quality
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.RotateLinear) = DeskewCheckCommandFlags.RotateLinear Then
            _rbLow.Checked = True
         End If
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.RotateResample) = DeskewCheckCommandFlags.RotateResample Then
            _rbMedium.Checked = True
         End If
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.RotateBicubic) = DeskewCheckCommandFlags.RotateBicubic Then
            _rbHigh.Checked = True
         End If

         ' Type
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.DocumentImage) = DeskewCheckCommandFlags.DocumentImage Then
            _rbTextOnly.Checked = True
         End If
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.DocumentAndPictures) = DeskewCheckCommandFlags.DocumentAndPictures Then
            _rbTextImages.Checked = True
         End If

         ' Speed
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.UseNormalRotate) = DeskewCheckCommandFlags.UseNormalRotate Then
            _rbNormal.Checked = True
         End If
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.UseHighSpeedRotate) = DeskewCheckCommandFlags.UseHighSpeedRotate Then
            _rbFast.Checked = True
         End If

         'Algorithm
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.DoNotUseCheckDeskew) = DeskewCheckCommandFlags.DoNotUseCheckDeskew Then
            _rbOrdinaryDeskew.Checked = True
         End If
         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.UseNormalCheckDeskew) = DeskewCheckCommandFlags.UseNormalCheckDeskew Then
            _rbBankCheckDeskew.Checked = True
         End If

         If (_DeskewCheckCommand.Flags And DeskewCheckCommandFlags.UseLineDetectionToDeskewCheck) = DeskewCheckCommandFlags.UseLineDetectionToDeskewCheck Then
            _rbLineDetectionDeskew.Checked = True
         End If
      End Sub

      Private Sub UpdateCommand()
         'DeskewCheckCommand.FillColor = new RasterColor(Color.Red);// _Color;
         _DeskewCheckCommand.Flags = CType(0, DeskewCheckCommandFlags)

         ' Fill
         If _cbFillExposedArea.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.FillExposedArea
         Else
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.DoNotFillExposedArea
         End If

         'Deskew Image
         If _rbDeskewImage.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.DeskewImage
         Else
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.ReturnAngleOnly
            ReturnAngleOnly = True
         End If

         ' Threshold
         If _cbThreshold.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.Threshold
         Else
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.NoThreshold
         End If

         ' Quality
         If _rbLow.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.RotateLinear
         ElseIf _rbMedium.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.RotateResample
         Else
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.RotateBicubic
         End If

         ' Type
         If _rbTextOnly.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.DocumentImage
         Else
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.DocumentAndPictures
         End If

         ' Speed
         If _rbNormal.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.UseNormalRotate
         Else
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.UseHighSpeedRotate
         End If

         'Algorithm
         If _rbOrdinaryDeskew.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.DoNotUseCheckDeskew
         End If
         If _rbBankCheckDeskew.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.UseNormalCheckDeskew
         End If
         If _rbLineDetectionDeskew.Checked Then
            _DeskewCheckCommand.Flags = _DeskewCheckCommand.Flags Or DeskewCheckCommandFlags.UseLineDetectionToDeskewCheck
         End If


      End Sub

      Private Sub _cbFillExposedArea_CheckedChanged(ByVal sender As Object, ByVal e As EventArgs) Handles _cbFillExposedArea.CheckedChanged
         _lblColor.Enabled = _cbFillExposedArea.Checked
      End Sub

      Private Sub lblColor_Click(ByVal sender As Object, ByVal e As EventArgs)
         Dim dlg As ColorDialog = New ColorDialog()
         dlg.Color = _lblColor.BackColor
         If dlg.ShowDialog() = DialogResult.OK Then
            _lblColor.BackColor = dlg.Color
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(dlg.Color)
         End If
      End Sub

      Private Sub _btnColor_Click(ByVal sender As Object, ByVal e As EventArgs) Handles _btnColor.Click
         Dim dlg As ColorDialog = New ColorDialog()
         dlg.AllowFullOpen = True
         dlg.AnyColor = True
         Dim res As DialogResult = dlg.ShowDialog(Me)
         If res = DialogResult.OK Then
            _Color = Leadtools.Demos.Converters.FromGdiPlusColor(dlg.Color)
         End If
         _lblColor.Refresh()
      End Sub

      Private Sub _lblColor_Paint(ByVal sender As Object, ByVal e As PaintEventArgs) Handles _lblColor.Paint
         e.Graphics.FillRectangle(New SolidBrush(Leadtools.Demos.Converters.ToGdiPlusColor(_Color)), _lblColor.ClientRectangle)
      End Sub
   End Class
End Namespace
